bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/discovery/HomeAssistantDiscovery.java

```diff
@@ -173,12 +175,31 @@ public void receivedMessage(ThingUID connectionBridge, MqttBrokerConnection conn
         HandlerConfiguration handlerConfig = new HandlerConfiguration(haID.baseTopic, topics);
         properties = handlerConfig.appendToProperties(properties);
         properties = config.appendToProperties(properties);
-        // First remove an already discovered thing with the same ID
-        thingRemoved(thingUID);
-        // Because we need the new properties map with the updated "components" list
-        thingDiscovered(DiscoveryResultBuilder.create(thingUID).withProperties(properties)
-                .withRepresentationProperty("objectid").withBridge(connectionBridge)
-                .withLabel(config.getThingName() + " (" + componentNames + ")").build());
+
+        synchronized (results) {
+            // Because we need the new properties map with the updated "components" list
+            results.put(thingUID.getAsString(),
+                    DiscoveryResultBuilder.create(thingUID).withProperties(properties)
+                            .withRepresentationProperty("objectid").withBridge(connectionBridge)
+                            .withLabel(config.getThingName() + " (" + componentNames + ")").build());
+        }
+    }
+
+    protected void publishResults() {
+        Collection<DiscoveryResult> localResults;
+
+        synchronized (results) {
+            localResults = new ArrayList<>(results.values());
+            results.clear();
+            componentsPerThingID.clear();
+        }
+        for (DiscoveryResult result : localResults) {
+            final ThingTypeUID typeID = result.getThingTypeUID();
+            ThingType type = typeProvider.derive(typeID, MqttBindingConstants.HOMEASSISTANT_MQTT_THING).build();
+            typeProvider.setThingTypeIfAbsent(typeID, type);
+
+            thingDiscovered(result);
```

Comment:
The original code removed things with the same `thingUID` before adding the result. Is that covered by some other logic?

Reply:
Yes. the whole change is about this.
The old code would discover one thing, when a topic is received.
Receiving a second topic for the same thing would change the discovered object and therefor the old object was removed before discovering the modified one.

The new code will wait for all topics to be received (via time-out, the time-out logic was already there). Then the discovered things will be published all at once. We will change the things in the waiting map, so we don't need to remove an already published thing because we did not publish it beforehand.